### PR TITLE
Elevation behavior

### PR DESCRIPTION
**Pull request by podraco**
### Description of Changes
Lastest set of updates for the Real time Elevation behavior with mockups added.

### Screenshots

https://user-images.githubusercontent.com/5509325/107318042-6c859d80-6a61-11eb-972d-815382a45ec3.mp4

https://user-images.githubusercontent.com/5509325/107318046-6db6ca80-6a61-11eb-940d-075f60ddc799.mp4

### CODE REVIEW

**Comment by HeaTTheatR** [comment]
@podraco 

Снимок экрана 2021-02-09 в 08 22 39

`demos/kitchen_sink/libs/baseclass/elevation.py`
`demos/kitchen_sink/libs/kv/elevation.kv`
`demos/kitchen_sink/screens_data.json`

**You didn't need to modify these files.**

These examples overload the `KitchenSink` application. Therefore, I want to put them in the documentation. For demonstration, it is enough that there is now:


https://user-images.githubusercontent.com/16930280/107319789-95814f00-6ab0-11eb-9103-4ad14f85db08.mov

**Comment by podraco** [comment]
Ok then. Should i delete all the modifications i did to the demo mods then?

**Comment by HeaTTheatR** [comment]
@podraco You just haven't done a `git pull` before making new changes :)

**Comment by podraco** [comment]
i'm reviewing the files...

**Comment by HeaTTheatR** [comment]
@podraco You need to do a `git pull`, make changes, and then do a `git push`

**Comment by podraco** [comment]
@HeaTTheatR , the master branch didn't allow me to work directly with it, i managed to do it by a fork repo that was up to date with the master branch when i started doing changes.
Let me review the files and ill discard whatever is not needed by my end.

**Comment by HeaTTheatR** [comment]
@podraco OK.

File: kivymd/uix/behaviors/elevation.py [hunk 1]
```diff
@@ -1145,37 +1145,25 @@ def __init__(self, **kwargs):
...
+                self._shadow = MDApp.get_running_app().theme_cls.rec_st_shadow
+                ratio = abs(ratio)
+                if ratio > 5:
+                    ratio = ratio * 22
+                else:
+                    ratio = ratio * 11.5
+                width = soft_width = self.width * 1.9
+                height = self.height + dp(ratio)
+                soft_height = (
+                    self.height + dp(ratio) + dp(self._elevation) * 0.5
```
Comment (HeaTTheatR) [inline kivymd/uix/behaviors/elevation.py:1166]: Are we back to old changes?

**Review by HeaTTheatR — CHANGES_REQUESTED** [review]

File: kivymd/uix/behaviors/elevation.py (hunk 1 above)
Comment (podraco) [reply kivymd/uix/behaviors/elevation.py:1166]: more like mockup shadows for bigger elements, 
the new shadows render the background dynamically, the old draws a mockup texture.
while the new shadows cast a better and 
stylized shadow and elevation. you can see how the shadows differ between the Fake**elevationbehavior and the Rectangular/circularElevarion behavior in the videos above.

Comment (podraco) [reply kivymd/uix/behaviors/elevation.py:1166]: While material IO ask for a dynamic shadow cast, stating that shadow should change with the widget, fake shadows are a great tool for most static widgets.

**Comment by HeaTTheatR** [comment]
@podraco Please clarify the status of this pull request at the moment, because I'm already confused :))

**Comment by podraco** [comment]
The pull request divides on 2 parts.

#### First part | Elevation.py.

These are the elevation classes that are currently available in the behavior

##### Real-time Shadows:
1. `RectangularElevationBehavior`: Draws a rectangular shadow based on the size of the widget that keeps consistency and keeps a natural-looking shadow
2. `CircularElevationBehavior`:  Draws a Circular shadow based on the size of the widget that keeps consistency and keeps a natural-looking shadow
3. `RoundedRectangularElevationBehavior`: Draws a rounded cornered rectangular shadow based on the size of the widget that keeps consistency and keeps a natural-looking shadow
4. `ObservableShadow`: it's  a proposal to allow faking partial shadows that I'm currently working on, this is supposed to draw only the part of the shadow that's on screen. - Still under development, but the name is already added for future use.


##### Fake Shadows:
1. `FakeRectangularElevationBehavior`: Fixed shadow based on preprocessed rectangular and square textures, the old way,
2. `FakeCircularElevationBehavior`: Fixed shadow based on preprocessed Circular textures, the old way,

The problem with this is that the shadow image and saturation won't match between a rectangular shadow and a square shadow. It's faster because uses stored shadows to draw, but it's not as clean as the Real-time elevation classes. 

This part of the pull request adds this classes back for developers to decide which elevation behavior their implementation requires.
While keeping the real-time shadows as the standard for every prefab widget.

This part of the pull request is intended for flexibility of kivymd.


#### Second part | Optimizations.
The second part of this pull request adds some optimizations, taking away the elevation control that some widgets took form the behavior, while adding a new section to the elevation behavior docs on how to properly animate the elevation of the widgets affected.

**Comment by HeaTTheatR** [comment]
@podraco But I still don't see any documentation on how to use ...

**Comment by podraco** [comment]
the idea is that they use them as any other behavior, all processing is internally done so they can just change the behavior they need, like 
``` py
class someclass(
    themable_behavior,
    shadow_behavior, # here they can choose between all our behaviors and they will work with the same standard API
    Specific_Color_Bheavior,
    ripple_behavior,
    higher_class_behavior.
):
```

**Comment by HeaTTheatR** [comment]
@podraco Unfortunately, I do not understand what this is about ...

**Comment by podraco** [comment]
i'm sorry i'm not sure how to express it more clearly.

**Comment by HeaTTheatR** [comment]
@podraco Me and users need specific code examples ...

**Comment by gregynard** [comment]
FYI the elevation for card is okay but the MDMDFloatingActionButton still the same. The resolution also is kinda different from the previous. the program is still sluggish there are some delays when you pressed the button and switch screens. 
Thanks!

**Comment by HeaTTheatR** [comment]
@gregynard Sorry, I know there are big problems with the new changes that related  this pull request. Unfortunately, the author of this pull request has not yet provided documentation on how to properly use these changes. Let's wait ...

**Comment by gregynard** [comment]
Oki oki. No problem. Thanks!

**Comment by podraco** [comment]
Sorry I was testing a theory. It seems that it worked :9
in the next pull request ill provide the necessary docs.

**Comment by HeaTTheatR** [comment]
@podraco 

Test with Pillow:


https://user-images.githubusercontent.com/16930280/107553682-0e35f780-6be6-11eb-9bd5-d4246b7d6698.mov

Test with Pillow-SIMD:


https://user-images.githubusercontent.com/16930280/107553756-2279f480-6be6-11eb-8f3e-52c11e226cc4.mov

Pillow-Simd - https://github.com/uploadcare/pillow-simd

**Comment by podraco** [comment]
never heard of that i was attepmting to use multithread to cast shadows in a secondary process, since the actual one blocks the GIL

**Comment by podraco** [comment]
how well does the corners works?

**Comment by gregynard** [comment]
Hi @podraco the elevations are now okay. But there are still some delays when switching screen or expanding a card. The mdfloating button also has no elevation. Thanks!

**Comment by podraco** [comment]
hii @gregynard 
the elevation documentation has been updated, i recommend you to check it up.
the new elevation classes uses real-time shadows, since the Material IO requires that the casted shadow changes with the widget, but, the FakeRectangular and FakeCircular elevation behaviors can improve the rendering time on square cards. and circular elements.
You can compile the latest docs directly with a local clone of the git repo or read the file directly.
as for the MDFloating button, i had no idea, let me check it up.

**Comment by gregynard** [comment]
Hi @podraco this is noted. Thank you!

**Comment by podraco** [comment]
got a fix for all buttons. but it seems the shadow of the MDToolbar MDfloating action button was broken all along...

**Comment by gregynard** [comment]
Hi @podraco thanks for this. I tried to follow the documentations I can now control the elevations.

**Comment by podraco** [comment]
it depends on how kivy is installed and used, as you can see, @HeaTTheatR  used Pillow-SIMD which shown a great performance improvement.
i as well am working on some performance solutions that will keep the shadows light weight and faster, but for now, I'm looking through the code, since some widgets take control from the behavior and duplicate some actions, like, redrawing the shadow multiple times outside the behavior or require a mock-up shadow instead of an RTC

**Comment by podraco** [comment]
If you're interested in some previews of the solutions, you're always welcome to checkup podraco/kivymd branches, I would appreciate some feedback before requesting a merge.

**Comment by gregynard** [comment]
Noted on this. Thanks thanks!

**Comment by podraco** [comment]
the last commits are now part of the master branch

**Pull request merged**